ext/LuxReactantExt.jl

```diff
@@ -0,0 +1,238 @@
+function LuxCore.initialparameters(rng::AbstractRNG, layer::Lux.ReactantLayer)
+    ps = layer.adaptor(LuxCore.initialparameters(rng, layer.layer))
+    layer.eltype_adaptor !== nothing && (ps = adapt(layer.eltype_adaptor, ps))
+    return __make_concrete_array(ps)
+end
+
+function LuxCore.initialstates(rng::AbstractRNG, layer::Lux.ReactantLayer)
+    st = LuxCore.initialstates(rng, layer.layer)
+    layer.eltype_adaptor !== nothing && (st = adapt(layer.eltype_adaptor, st))
+    return __make_concrete_array(st)
+end
+
+function (l::Lux.ReactantLayer{FST, T})(x, ps, st::NamedTuple) where {FST, T}
+    l.eltype_adaptor !== nothing && (x = adapt(l.eltype_adaptor, x))
+
+    # XLARuntimeError is not great, so check and terminate early if needed
+    input_structure = fmapstructure(Lux.__size, x)
+    if l.input_structure != input_structure
+        throw(DimensionMismatch(lazy"Input structure mismatch. Expected $(l.input_structure), got $(input_structure)."))
+    end
+
+    # TODO: For non array inputs this we make the eltype uniform which might not be
+    #       desirable. We should handle those cases with `fmap`
+    if T != Lux.__recursive_eltype(x)
+        @warn """
+        `Reactant.compile` was called with input eltype $(T) but the current input eltype \
+        is $(Lux.__recursive_eltype(x)). This might lead to unexpected behavior.
+
+        We will convert the input to $(T) and continue. If you want to avoid this, please \
+        recompile the model with the correct input eltype.
+        """ maxlog=1
+        x = adapt(LuxEltypeAdaptor{T}(), x)
+    end
+
+    return Lux.__apply_reactant(l, x, ps, st)
+end
+
+# This is the ideal case where all the types match correctly.
+# Input Type mispatches should not happen here, they should be handled before this function
+# is called.
+# If `st` mismatch happens then user really messed something up. can't do anything about it.
+@inline function Lux.__apply_reactant(
+        l::Lux.ReactantLayer{FST, T, inType}, x::inType, ps, st) where {FST, T, inType}
+    return Lux.__apply_reactant(l, __make_concrete_array(x), ps, st)
+end
+
+@inline function Lux.__apply_reactant(
+        l::Lux.ReactantLayer{FST, T, inType, inCType, psType, stType}, x::inCType,
+        ps::psType, st::stType) where {FST, T, inType, inCType, psType, stType}
+    csmodel = Lux.StatefulLuxLayer{FST}(l.clayer, ps, st)
+    return Lux.__apply_reactant(l, csmodel, x), ifelse(FST, csmodel.st, csmodel.st_any)
+end
+
+# Parameter type mismatch. This might be too common so try to handle it gracefully.
+@inline function Lux.__apply_reactant(
+        l::Lux.ReactantLayer{FST, T, inType, inCType, psType, stType}, x::inCType,
+        ps::psType2, st::stType) where {FST, T, inType, inCType, psType, psType2, stType}
+    ps = __try_similar_structure(Lux.__named_tuple(ps), l.concrete_ps)
+    ps = l.adaptor(ps)
+    l.eltype_adaptor !== nothing && (ps = adapt(l.eltype_adaptor, ps))
+    ps = __make_concrete_array(ps)
```

Comment:
can this be set up where the ps are already reactant arrays so we don't need to call __make_concrete_array

Reply:
[hope would be avoiding data shuffling, especially cpu<->gpu of the whole model]

Reply:
This should not be hit in most cases. I wasn't able to compile a ComponentArrays based version yet (needs a `make_tracer` overload), so it is a temporary solution.

The correct use of this should hit L163